### ISSUE TITLE
.NET 8: Enable private NuGet dependencies for alternative licensing

### ISSUE THREAD

**Issue by richlander** [dotnet/sdk#26286]
The ImageSharp project is considering a [new licensing scheme](https://github.com/SixLabors/ImageSharp/discussions/2151). The license language discusses a bunch about the nature of dependencies. It seems like private NuGet dependencies (for implementation-only dependencies (no API exposure) could be an important tool for enabling projects to establish alternate licensing approaches. At present, that's difficult.

It would be good to validate the assumption that private NuGet dependencies would help. It seems like they would.

After that, it would be good to consider this feature for .NET 8. It has other benefits as well. It is a subset of the functionality proposed in: https://github.com/dotnet/designs/pull/242.

@clairernovotny @mhutch @JimBobSquarePants

**Comment by Sergio0694** [comment]
This looks very promising! I wonder if it's essentially the same as https://github.com/NuGet/Home/issues/10375, or whether it at least overlaps with that? We've wanted something like this for years in the Windows Community Toolkit, and I could use this on my own as well 😄

**Comment by richlander** [comment]
Yes, same (or same enough). Let's leave them both open for a bit. We can close this one once the team goes through planning. Mine is pushing (quite obviously) on one specific scenario to help motivate the experience.

**Comment by JimBobSquarePants** [comment]
In my proposed license change I'm not treating a direct dependency as "You call this API", rather, "You explicitly installed the package or included the source" it so I'm not sure of the relevance here. However I see the relevance in more traditional direct vs transitive dependency definition.

**Comment by richlander** [comment]
I think we're talking about the same thing, @JimBobSquarePants.

Private NuGet dependencies are an answer to the problem you are mentioning. They make it so that you have to explicitly install the package to use it. If another package is using ImageSharp as an implementation-only dependency, the app user doesn't see it.

If you don't care about this topic, then that means the more popular ImageSharp gets, the less your test passes "you explicitly installed the package". For example, Json.NET is pretty much everywhere. You can install tons of other packages, and you can end up being able to call JSON.NET APIs w/o explicitly installing the package. Is that a fine outcome from your point of view?

**Comment by JimBobSquarePants** [comment]
Ah riiiight.... OK Yeah. Private NuGet dependencies solve the transitive bad actor problem really well!

> Is that a fine outcome from your point of view?

Well.... fine is subjective. The way the proposed license works is that if the party that took the direct dependency qualifies for the Apache License then transitive consumption falls under that license. My Commercial License would grant similar freedoms so in real terms I wouldn't have means to stop consumption due to the non-virility of the Split License.

I'm stuck with the problem in that, because this is an established library with a massive OSS userbase with permissive licenses, I have to avoid license virility. I simply can't cut people off so I have to work with loopholes. 

For others starting fresh this is an awesome tool to have.

**Comment by richlander** [comment]
It's possibly slightly better than that. Imagine there were n very popular packages on NuGet that used ImageSharp. You could talk to them and ask them to use this feature. Meaning, you can do some repair on past usage, too.

**Comment by zivkan** [comment]
What does "private dependency" actually mean? I've only skimmed over this issue, so I'm sorry if I missed it.

Does the existing NuGet feature `PrivateAssets="compile"` already meet the objective?  https://github.com/dotnet/roslyn-sdk/blob/df51b014fcb8c9e667a0b09a000f2050615f7afb/src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.Analyzer.Testing/Microsoft.CodeAnalysis.Analyzer.Testing.csproj?q=PrivateAssets%3D%22compile%22#L58-L65

**Comment by richlander** [comment]
Great question, but I believe that feature does something different. AFAIK, that's a consumption-based feature. It is declaring whether you want the NuGet asset copied to your `bin` folder or used for a more exclusive purpose. The private dependency suggestion discussed in this issue is a production-based feature. It specifies whether a given dependency should be publicly available within the NuGet graph.

Did I get that right @mhutch @clairernovotny?

**Comment by zivkan** [comment]
https://docs.microsoft.com/en-us/nuget/consume-packages/package-references-in-project-files#controlling-dependency-assets

My summary for only the relevant parts:

metadata|in current project|flow via PackageReference
---|---|---
IncludeAssets| ✔️ | ✔️ 
PrivateAssets|  ✔️ | ❌ 
ExcludeAssets| ❌ | ❌ 

Asset type|description
--|--
compile|`lib/<tfm>/*.dll` or `ref/<tfm>/*.dll` passed to compiler
runtime|`lib/<tfm>/*dll` or `runtimes/<it's complicated>/*.dll` copied to `bin`
others|not relevant to this discussion

Therefore:

scenario|what to use
--|--
Prevent NuGet assets being copied into bin folder|`ExcludeAssets="all"` or `ExcludeAssets="compile"` if you want to keep non-dll assets, and exclude only the dlls
Don't allow consumers of my package to use my dependencies, unless they directly reference those packages themselves|`PrivateAssets="compile"

**Comment by richlander** [comment]
Interesting. I didn't know that. I'm interested to hear if there is nay nuance from the NuGet PMs. If we have the feature already, that's great.

**Comment by zivkan** [comment]
I just remembered that I documented this a while ago: https://docs.microsoft.com/en-us/nuget/create-packages/select-assemblies-referenced-by-projects#recommended-one-assembly-per-package

**Comment by richlander** [comment]
Cool. I just looked at https://www.nuget.org/packages/Microsoft.ServiceHub.Client/3.0.3078 in NPE.

I also added that library as PackageReference. I can still see this ns in IntelliSense: `Microsoft.VisualStudio.Threading`. Is that expected?

**Comment by richlander** [comment]
You can look at the nuspec in NPE. I did.

**Comment by inoa-pmattos** [comment]
Is there any news on this feature?

**Comment by JimBobSquarePants** [comment]
I've reread this conversion a few times now after a long break and I'm struggling to understand parts. 

The requirement is thus:

We need a way for a package (A) to reference another package (B) without exposing the public API of the referenced package (B). The user should be able to simply install package (A), write some code using the API for that package and be able to run the code without issue. Ideally there should be a single compiled binary within package (A) containing the code from (A) and (B). 

Are we saying @zivkan that this is already possible because I have not found this to be so using combinations of `(Private|Exclude|Include)Assets`

This feature is not only important for licensing, but also critical for security. Currently libraries will embed ImageSharp in their own libraries by using the source directly and changing public members to internal. This removes the ability to track CVEs and exposes users to risks that they have no way to be aware of.

**Comment by richlander** [comment]
@JonDouglas
